### ISSUE TITLE
Potential inconsistencies between paper and implementation in '_score()' and '_similarity_matrix()'

### ISSUE THREAD

**Issue by OmerFarukTal** [Kim-Minseon/TARO#1]
Hi, and thank you for releasing the code for “Effective Targeted Attacks for Adversarial Self-Supervised Learning (TARO)”!
While studying the implementation, I noticed a few points where the current code seems to differ from the paper’s equations (Eqs. 10–11):
Sentropy(x, x′) = p′/τ log (p′/τ ), Ssimilarity(x, x′) = e / |e|_2 * e' / |e'|_2 . (10) 
STARO(x, x′) = Sentropy + Ssimilarity. (11)

I’m writing this issue to confirm whether these were intentional design choices or implementation oversights.

## Cosine similarity scope (_similarity_matrix)

Paper:
> The cosine similarity is calculated between features of base images and candidate images in the differently augmented batch.

Code:

`similarity_value = torch.mm(outputs_norm, outputs_norm.transpose(0,1).detach())`

This computes intra-batch similarity (e1 × e1ᵀ or e2 × e2ᵀ), not cross-batch similarity (e1 × e2ᵀ).
Should the similarity instead be computed between e1 (base) and e2 (augmented) to match the paper?

Example of the alternative formulation:

```
cross_sim = torch.mm(
    F.normalize(e1, dim=1),
    F.normalize(e2, dim=1).t()
)
```

## Entropy sign and definition

Paper (Eq. 10):

𝑆entropy(𝑥, 𝑥′) = (𝑝′/𝜏) log(𝑝′/𝜏)

Code:

`entropy_value = (soft_output * log_soft_output).sum(1)`

This produces positive negative-entropy values (∑p log p), not true entropy (-∑p log p).
Was the missing negative sign intentional for ranking purposes, or should this term be negated?

## Ambiguity in entropy definition (self- vs cross-entropy)

Equation (10) defines

𝑆entropy(𝑥, 𝑥′) = (𝑝′/𝜏) log(𝑝′/𝜏)

which depends only on 𝑝′
However, the text says the score aims to select an instance 𝑥′ that is “confused with” the given image 𝑥, suggesting a potential cross-entropy term

𝑆entropy(𝑥, 𝑥′) = (𝑝′/𝜏) log(𝑝/𝜏)

that would directly compare the two predictions 𝑝 and 𝑝′.

Could you please clarify whether the intended definition of Sentropy is:

self-entropy (as in the current code), or cross-entropy between 𝑝 and 𝑝′ (as implied by the descriptive text)?

## Missing temperature term (τ)

The equations explicitly include a temperature τ, but the current implementation computes:

`soft_output = F.softmax(outputs.detach(), dim=1)`

without dividing logits by τ.
Should outputs be scaled as outputs / tau? If so what is the value of τ, as far as I could detect it is not mentioned in the paper.

## Normalization and diagonal handling

The similarity matrix currently adds:

```
similarity_value = _similarity_matrix(B, feature.view(B, -1))
eye = torch.eye(len(similarity_value)).cuda()*-1
similarity_value = similarity_value +eye
similarity_value = (similarity_value+1.0)/2.0
```

This sets self-similarities to ~0.5 rather than removing them.
Would masking or zeroing the diagonal be more consistent with the intent of excluding self-pairs?

# 🙏 Question

Could you please confirm:

Which version of '_score()' was used to produce the results in the paper?

Were these simplifications intentional for efficiency/stability, or do they reflect an earlier prototype?

Thanks again for sharing your work — this clarification would really help future reproductions and extensions!
